File: expensereport-lua/ExpenseReport.lua
ExpenseType = {
    DINNER = 1,
    BREAKFAST = 2,
    CAR_RENTAL = 3,
}

Expense = { type, amount }

function Expense:new(type, amount)
    o = {}
    setmetatable(o, self)
    self.__index = self
    o.type = type
    o.amount = amount
    return o
end

function printReport(expenses)
    mealExpenses = 0
    total = 0
    print("Expenses: " .. os.date())
    for i, expense in ipairs(expenses) do
        if expense.type == ExpenseType.DINNER or expense.type == ExpenseType.BREAKFAST then
            mealExpenses = mealExpenses + expense.amount
        end
        expenseName = ""
        if expense.type == ExpenseType.DINNER then expenseName = "Dinner" end
        if expense.type == ExpenseType.BREAKFAST then expenseName = "Breakfast" end
        if expense.type == ExpenseType.CAR_RENTAL then expenseName = "Car Rental" end
        mealOverExpensesMarker = (expense.type == ExpenseType.DINNER and expense.amount > 5000 or expense.type == ExpenseType.BREAKFAST and expense.amount > 1000) and "X" or " "
        print(expenseName .. "\t" .. expense.amount .. "\t" .. mealOverExpensesMarker)
        total = total + expense.amount
    end
    print("Meal Expenses: " .. mealExpenses)
    print("Total Expenses: " .. total)
end

expenses = {
    Expense:new(ExpenseType.DINNER, 5000),
    Expense:new(ExpenseType.DINNER, 5001),
    Expense:new(ExpenseType.BREAKFAST, 1000),
    Expense:new(ExpenseType.BREAKFAST, 1001),
    Expense:new(ExpenseType.CAR_RENTAL, 4),
}

printReport(expenses)

File: expensereport-scala/src/main/scala/com/nelkinda/training/ExpenseReport.scala
package com.nelkinda.training

import java.util.Date

enum ExpenseType:
  case DINNER, BREAKFAST, CAR_RENTAL

class Expense(
    val `type`: ExpenseType,
    val amount: Int
)

class ExpenseReport:
  def printReport(expenses: List[Expense]): Unit =
    var total = 0
    var mealExpenses = 0

    println("Expenses " + new Date())

    for (expense <- expenses)
      if (expense.`type` == ExpenseType.DINNER || expense.`type` == ExpenseType.BREAKFAST)
        mealExpenses += expense.amount

      var expenseName = ""
      expense.`type` match
        case ExpenseType.DINNER =>
          expenseName = "Dinner"
        case ExpenseType.BREAKFAST =>
          expenseName = "Breakfast"
        case ExpenseType.CAR_RENTAL =>
          expenseName = "Car Rental"

      val mealOverExpensesMarker =
        if (
          expense.`type` == ExpenseType.DINNER && expense.amount > 5000 ||
          expense.`type` == ExpenseType.BREAKFAST && expense.amount > 1000
        ) "X"
        else " "

      println(expenseName + "\t" + expense.amount + "\t" + mealOverExpensesMarker)

      total += expense.amount

    println("Meal expenses: " + mealExpenses)
    println("Total expenses: " + total)
